src/box/sql/resolve.c

```diff
@@ -130,34 +130,48 @@ resolveAlias(struct ExprList *pEList, int iCol, struct Expr *pExpr,
 	sql_xfree(pDup);
 }
 
-/*
- * Return TRUE if the name zCol occurs anywhere in the USING clause.
- *
- * Return FALSE if the USING clause is NULL or if it does not contain
- * zCol.
+/**
+ * Return TRUE if the name occurs anywhere in the USING clause.
+ * Return FALSE if the USING clause is NULL or if it does not contain the name.
  */
-static int
-nameInUsingClause(IdList * pUsing, const char *zCol)
+static bool
+nameInUsingClause(struct IdList *pUsing, const char *zCol, const char *old_col)
 {
 	if (pUsing) {
-		int k;
-		for (k = 0; k < pUsing->nId; k++) {
+		for (int k = 0; k < pUsing->nId; k++) {
 			if (strcmp(pUsing->a[k].zName, zCol) == 0)
-				return 1;
+				return true;
+		}
+		for (int i = 0; i < pUsing->nId; i++) {
+			if (strcmp(pUsing->a[i].zName, old_col) == 0)
+				return true;
 		}
 	}
-	return 0;
+	return false;
 }
 
 bool
-sqlMatchSpanName(const struct Token *span, const char *col, const char *tab)
+sqlMatchSpanName(const struct Token *span, const char *col, const char *tab,
+		 const char *old_col, const char *old_tab)
 {
```

Comment:
Looking at that and how you did the double lookups by allocating the legacy name on each failed lookup, I am wondering if there was another option.

Have you thought about introducing a `struct sql_name` or something alike? That would store 2 names, one original and one normalized. The normalized would be optional. And we would just pass it everywhere instead of `Token` where it is used as an identifier. If the normalized name is set, then you use it. And otherwise you don't.

It could also help places like this where which now got quite ugly with all those proper + "old" names. With `sql_name` you would pass this single object instead of 2 explicit version arguments of the name. And it would allow not to allocate the normalized name on each failed lookup of the original name.

Reply:
I though about adding `struct sql_name`, but found it quote inconvenient. Partly because of too many structures.

I replaced `has_*_lookup` flags by `char *legacy_name;` field in `struct ExprList_item`, `struct IdList_item` and `struct SrcList_item`. I tried to do the same in `struct Expr` but it is quite troublesome because ofvarious compressions of `struct Expr`.

Reply:
Too many structures? Do we get billed for each new struct 😁? The current solution looks quite clumsy, like I said above, and I think storing those 2 names in a struct would make the code cleaner. But up to you in this case.

Reply:
Well, I definitely agree with @Gerold103 here, that it looks more convenient to encapsulate all double lookup machinery. However, I have no strong opinion here regarding the way it should be done (I have no such a deep knowledge in SQL engine at the moment). So, if @ImeevMA has any ideas on making this code cleaner and readable today, feel free to do it. Otherwise, we can postpone such kind of refactoring until better times.